src/Orleans/Messaging/IncomingMessageBuffer.cs

```diff
@@ -1,44 +1,329 @@
+        protected int receiveOffset;
+        protected int decodeOffset;
 
-        private int headerLength;
-        private int bodyLength;
+        protected readonly bool supportForwarding;
+        protected Logger Log;
 
-        private int receiveOffset;
-        private int decodeOffset;
+        private readonly MessagePrefixHolder prefixHolder = new MessagePrefixHolder();
 
-        private readonly bool supportForwarding;
-        private Logger Log;
-
-        public IncomingMessageBuffer(Logger logger, bool supportForwarding = false, int receiveBufferSize = DEFAULT_RECEIVE_BUFFER_SIZE, int maxSustainedReceiveBufferSize = DEFAULT_MAX_SUSTAINED_RECEIVE_BUFFER_SIZE)
+        protected IncomingMessageBuffer(Logger logger,
+            bool supportForwarding = false,
+            int receiveBufferSize = DEFAULT_RECEIVE_BUFFER_SIZE,
+            int maxSustainedReceiveBufferSize = DEFAULT_MAX_SUSTAINED_RECEIVE_BUFFER_SIZE)
         {
             Log = logger;
             this.supportForwarding = supportForwarding;
             currentBufferSize = receiveBufferSize;
             maxSustainedBufferSize = maxSustainedReceiveBufferSize;
             lengthBuffer = new byte[Message.LENGTH_HEADER_SIZE];
-            readBuffer = BufferPool.GlobalPool.GetMultiBuffer(currentBufferSize);
             receiveOffset = 0;
             decodeOffset = 0;
             headerLength = 0;
             bodyLength = 0;
         }
 
+        public IncomingMessageBuffer(
+            Logger logger,
+         List<ArraySegment<byte>> readBuf,
+         bool supportForwarding = false,
+         int receiveBufferSize = DEFAULT_RECEIVE_BUFFER_SIZE,
+         int maxSustainedReceiveBufferSize = DEFAULT_MAX_SUSTAINED_RECEIVE_BUFFER_SIZE)
+            : this(logger, supportForwarding, receiveBufferSize, maxSustainedReceiveBufferSize)
+        {
+            readBuffer = readBuf;
+        }
+
+        private int AvailableReadLength => prefixHolder.Count + receiveOffset - decodeOffset;
+
+        public virtual void UpdateReceivedData(int bytesRead)
+        {
+            receiveOffset = bytesRead;
+
+            // after each recieve buffer gets filed from the begining
+            decodeOffset = 0;
+
+            // Opportunistic reset of the prefix holder
+            if (prefixHolder.HasPrefix && decodeOffset >= prefixHolder.Count)
+            {
+                prefixHolder.Reset();
+            }
+        }
+
+        public void Reset()
+        {
+            receiveOffset = 0;
+            decodeOffset = 0;
+            headerLength = 0;
+            bodyLength = 0;
+        }
+
+        public virtual bool TryDecodeMessage(out Message msg)
+        {
+            msg = null;
+            if (decodeOffset == receiveOffset)
+            {
+                return false;
+            }
+
+            // Is there enough read into the buffer to continue (at least read the lengths?)
+            if (TryHandleMessageLengthPrefix()) return false;
```

Comment:
> I've avoided unnecessary check. 

What unnecessary check?
headerLenght== 0 || bodyLength ==0?  This is a cheaper check than TryHandleMessageLengthPrefix.  In the most common case, both checks are made anyway.  In the less common case, only headerLenght== 0 || bodyLength (the cheaper check) should be necessary.

_correctness_
Improbable but not impossible scenario.  What happens if a read gets 8 bytes, (we read lengths but not  header or body and nothing in holder), then header and body are read in next loop but they are less than 8?  Say they are 6 bytes?

AvailableReadLenght (0 + 6 – 0) < readLength (8)

In this case the header and body are there, but we don’t process them because we are looking for the lengths size.

If we’re not looking for lengths, because they’ve already been read, why are we checking if there is enough data to read them?  While this scenario is not likely, the logic is technically incorrect.

Am I missing something?